### PR TITLE
Add the curve25519 curve

### LINKED ISSUE #115: Plan to handle curve25519 and ed25519
## Summary

Our new framework explicitly defines the Montgomery curve corresponding to a twisted Edwards curve. 

This is, however, not unifying curve25519 and ed25519.

Ed25519 is a twisted Edwards curve, and is "birationally equivalent" to curve25519, but the mapping of A/B to a/d as well as u/v to x/y is different from standard formulas.

This is done with a good cause that ed25519 also needs to use small parameters to facilitate computation, so such adjustment is actually justified.

This, however, seems to suggest that we need to define curve25519 and ed25519 as two separate curves here. These two curves can additionally implement more functions (not covered by traits) for birationally equivalence, if necessary.

## Proposal

- have both curve25519 and ed25519
- (optionally) birationally equivalence

____

#### For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

This PR is related to #121 and #115. Note that curve25519 is a Montgomery curve, but its corresponding TE curve is not ed25519.

Closes #115.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: curve25519/src/curves/mod.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// We want to emphasize that this twisted Edwards curve is not ed25519.
+impl TECurveConfig for EdwardsParameters {
+    /// COEFF_A = 486664
+    const COEFF_A: Fq = MontFp!("486664");
+
+    /// COEFF_D = 486660
+    const COEFF_D: Fq = MontFp!("486660");
+
+    /// Generated randomly
+    const GENERATOR: EdwardsAffine = EdwardsAffine::new_unchecked(GENERATOR_X, GENERATOR_Y);
```
Comment (Pratyush) [inline curve25519/src/curves/mod.rs:40]: We should use the standard generator here, no?

File: curve25519/src/curves/mod.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+#[cfg(test)]
+mod tests;
+
+pub type EdwardsAffine = Affine<EdwardsParameters>;
+pub type EdwardsProjective = Projective<EdwardsParameters>;
+pub type NonZeroMontgomeryAffine = MontgomeryAffine<EdwardsParameters>;
+
+#[derive(Clone, Default, PartialEq, Eq)]
+pub struct EdwardsParameters;
```
Comment (Pratyush) [inline curve25519/src/curves/mod.rs:16]: Maybe we should rename this to `Curve25519Config`?

**Comment by Pratyush** [comment]
What does it mean that the "corresponding TE curve is not ed25519"?

**Comment by Pratyush** [comment]
This looks good, but it would be nice (either in this PR, or in a followup PR) to use specialized reductions for the special form of the base field modulus.

File: curve25519/src/curves/mod.rs (hunk 1 above)
Comment (weikengchen) [reply curve25519/src/curves/mod.rs:40]: This is the Edwards equivalence of the standard generator (x=9) of the Montgomery curve:
(0x09, 0x20ae19a1b8a086b4e01edd2c7748d14c923d4d7e6d7c61b229e9c5a27eced3d9)
=>
(38213832894368730265794714087330135568483813637251082400757400312561599933396, 46316835694926478169428394003475163141307993866256225615783033603165251855960)

as shown in the script

File: curve25519/src/curves/mod.rs (hunk 2 above)
Comment (weikengchen) [reply curve25519/src/curves/mod.rs]: Renamed.

**Comment by weikengchen** [comment]
> What does it mean that the "corresponding TE curve is not ed25519"?

Updated.

**Comment by weikengchen** [comment]
@arkworks-rs/algebra This one is ready for review.

File: curve25519/src/curves/mod.rs (hunk 1 above)
Comment (Pratyush) [reply curve25519/src/curves/mod.rs:41]: Ok, could you update the comment (which currently says `Generated randomly`)  in that case?

File: curve25519/src/curves/mod.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    /// COEFF_A = 486664
+    const COEFF_A: Fq = MontFp!("486664");
+
+    /// COEFF_D = 486660
+    const COEFF_D: Fq = MontFp!("486660");
+
+    /// Standard generators from https://neuromancer.sk/std/other/Curve25519.
+    /// The Montgomery form is (0x09, 0x20ae19a1b8a086b4e01edd2c7748d14c923d4d7e6d7c61b229e9c5a27eced3d9).
+    /// The twisted Edwards form is 
+(38213832894368730265794714087330135568483813637251082400757400312561599933396, 46316835694926478169428394003475163141307993866256225615783033603165251855960)
```
Comment (weikengchen) [inline curve25519/src/curves/mod.rs:43]:
```suggestion
/// (38213832894368730265794714087330135568483813637251082400757400312561599933396, 46316835694926478169428394003475163141307993866256225615783033603165251855960)
```

**Comment by weikengchen** [comment]
Fixed the comments here as well.

**Review by Pratyush — APPROVED** [review]

File: curve25519/src/curves/mod.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+// We want to emphasize that this twisted Edwards curve is not ed25519.
+// Ed25519 has COEFF_A = -1 and COEFF_D = -121665 / 121666.
+impl TECurveConfig for Curve25519Config {
+    /// COEFF_A = 486664
+    const COEFF_A: Fq = MontFp!("486664");
+
+    /// COEFF_D = 486660
+    const COEFF_D: Fq = MontFp!("486660");
+
+    /// Standard generators from https://neuromancer.sk/std/other/Curve25519.
```
Comment (weikengchen) [inline curve25519/src/curves/mod.rs:40]:
```suggestion
    /// Standard generators from <https://neuromancer.sk/std/other/Curve25519>.
```

**Pull request merged**